File: slabc/R/GKDR.R
GKDR <- function(G, obstemp, theta, focuspara,ss_set) {
  Np = 20000
  Nt = 30000
  N = 2000
  np = 3
  weight_k = 1
  c = 0.75
  
  CV_num = 1000
  # generate training data
   train = simRicker_gkdr(Np,ss_set)
   test = simRicker_gkdr(Nt,ss_set)
  # save(train, test, file='gkdr_test_train.RData')
  #load('gkdr_test_train.RData')
  ##choose set of original summary statistics
  #obstemp=obstemp$E0#1:E0,2:E1,3:E2
  
  test.sy=t(test$sy)
  train.sy = t(train$sy)
  # test.sy.std=apply(test.sy,2,sd)
  train.sy.std=apply(train.sy,2,sd)
  
  train.sy.mean=apply(train.sy,2,mean)
  
  test.sy=sweep(test.sy,2,train.sy.mean,FUN="-")
  test.sy=sweep(test.sy,2,train.sy.std,FUN="/")
  train.sy=sweep(train.sy,2,train.sy.mean,FUN="-")
  train.sy=sweep(train.sy,2,train.sy.std,FUN="/")
  ##
  P = length(obstemp)
  obstemp=(obstemp-train.sy.mean)/train.sy.std
  SumStatDat_v = rep(obstemp, each = Np)
  SumStatDat = matrix(SumStatDat_v, nrow = Np)
  SumStatDatt_v = rep(obstemp, each = Nt)
  SumStatDat_t = matrix(SumStatDatt_v, nrow = Nt)
  # prepare data

  thetas = cbind(train$log_r, train$log_sig_e, train$phi)
  thetast = cbind(test$log_r, test$log_sig_e, test$phi)
  
  sprintf('#sample = %d, dim of X = %d, effective dim = %d\n', N, P, G)
  candx = c(0.25, 0.5, 0.75, 1,2)  # candidates for CV
  candy = c(0.25, 0.5, 1,2)
  eps = 0.00001;
  err_tbl = vector(mode = "numeric",
                   length = (length(candx) * length(candy) * length(eps)))
  err_tbl2 = vector(mode = "numeric",
                   length = (length(candx) * length(candy) * length(eps)))
  err_tbl3 = vector(mode = "numeric",
                   length = (length(candx) * length(candy) * length(eps)))
  #sorting to get bandwidth and training data
  rsse_b = rowMeans((SumStatDat - train.sy) * (SumStatDat - train.sy))
  rsse_t = sort(rsse_b, index.return = TRUE)
  
  Xm = train.sy[rsse_t$ix,]
  Ym = thetas[rsse_t$ix,]
  Y = Ym[1:N,]
  X = Xm[1:N,]
  rtt = rsse_t$x[1:N]
  rsseup = rtt[N]
  W = c * (1 - (rtt / rsseup) ^ 2)
  # Xsd=apply(X,2,sd)
  # Ysd=apply(Y,2,sd)
  # X=sweep(X, 2, obs_sd, FUN="/")
  # Y=sweep(Y, 2, Ysd, FUN="/")
  # X=scale(X, center = FALSE, scale = TRUE)
   Y=scale(Y, center = TRUE, scale = TRUE)
  # train.sy=sweep(train.sy,2,obs_sd,FUN="/")
  # SumStatDat_t=sweep(SumStatDat_t,2,obs_sd,FUN="/")
  sgx0 = MedianDist(X)
  sgy0 = MedianDist(Y)
 # X=X*W
  # parallel validation
  sgx_v = sgx0 * candx
  sgy_v = sgy0 * candy
  tt1 = length(candx) * length(candy) * length(eps)
  tt2 = length(candy) * length(eps)
  for (hhh in 1:tt1) {
    opth = ceiling(hhh / tt2)
    sgx = sgx_v[opth]
    rr = hhh - (opth - 1) * tt2
    optk = ceiling(rr / length(eps))
    sgy = sgy_v[optk]
    opte=rr-(optk-1)*length(eps)

    EPS = eps[opte]
    time.each=proc.time()
    KD = KernelDeriv(X, Y, K, sgx, sgy, EPS, W)
    time.eachend=proc.time()-time.each
    #cat(sprintf("\n time for KernelD in %d round is %f\n",hhh,time.eachend))
    K=8#nScree(x=KD$values)$Components$nkaiser
    B=KD$vectors[,1:K]
    rsse_b = rowMeans((SumStatDat_t%*%B - test.sy%*%B)^2)
    #reesb_rej=rowMeans((SumStatDat_t - test.sy)^2)
    rsse_t = sort(rsse_b, index.return = TRUE)
    #rsse_t_rej = sort(reesb_rej, index.return = TRUE)
    Yt = thetast[rsse_t$ix,]
    #Ytt = thetast[rsse_t_rej$ix,]
    Yt_v = colMeans(Yt[1:100,])
    Yt_v2 = colMeans(Yt[1:10,])
    Yt_v3 = colMeans(Yt[1:5,])
    #Ytt_v = colMeans(Ytt[1:20,])
    err_tbl[hhh] = sum((Yt_v-theta)^2)
    err_tbl2[hhh] = sum((Yt_v2-theta)^2)
    err_tbl3[hhh] = sum((Yt_v3-theta)^2)

    #errr=sum((Ytt_v-theta)^2)
  }
  midx = which.min(err_tbl)
  midx2 = which.min(err_tbl2)
  midx3 = which.min(err_tbl3)
  opth = ceiling(midx3 / tt2)
  sgx = sgx_v[opth]
  rr = midx3 - (opth - 1) * tt2
  optk = ceiling(rr / length(eps))
  sgy = sgy_v[optk]
  opte=rr-(optk-1)*length(eps)

  #set up X,Y and weight again as opimum value
  sgx = sgx0*candx[opth]
  sgy = sgy0*candy[optk]
  EPS = eps[opte]

  KD = KernelDeriv(X, Y, K, sgx, sgy, EPS, W)
  K=8#nScree(x=KD$values)$Components$nkaiser
  para_gkdr = c(sgx, sgy, EPS)
  B=KD$vectors[,1:K]
  
  cat(sprintf("Npop: %d,N: %d", Np, N))
  cat(sprintf(
    "/n K= %d sgx=%f,sgy=%f,eps:%f,Yfocus:%d\n",
    K,
    sgx,
    sgy,
    EPS,
    focuspara
  ))
  return(B)
  }


MedianDist <- function (X) {
  N = length(X[, 1])
  ab = X %*% t(X)
  aa = diag(ab)
  X=matrix(rep(c(aa),times=N),nrow = N)
  Dx=X+t(X)-2*ab
  Dx = Dx - diag(diag(Dx))
  dx = c(Dx)
  dx = dx[dx != 0]
  s = sqrt(median(dx))
}

KernelDeriv <- function(X, Y, K, sgx, sgy, EPS, W){
  dims=dim(X)
  N=dims[1]
  M=dims[2]
  sx2=sgx^2
  sy2=sgy^2
  
  ab=X%*%t(X)
  aa=matrix(rep((diag(ab)),times=N),nrow = N)
  xx=aa+t(aa)-2*ab
  xx[xx<0]=0
  Kx=exp(-xx*(sx2^(-1)))
  
  ab=Y%*%t(Y)
  aa=matrix(rep((diag(ab)),times=N),nrow = N)
  yy=aa+t(aa)-2*ab
  yy[yy<0]=0
  Ky=exp(-yy*(sy2^(-1))) 
  
  F1=solve(Kx+N*EPS*diag(N))%*%Ky%*%solve(Kx+N*EPS*diag(N))
  Mt=array(0,dim=c(M,M,N))
  Xmt=matrix(rep(X,each=N,times=1),nrow = N)
  dim(Xmt)=c(N,N,M)
  Xm=(aperm(Xmt,c(2,1,3))-Xmt)/sx2
  Kxm=rep(Kx,times=M)
  dim(Kxm)=c(N,N,M)
  H=matrix(Xm)*matrix(Kxm)
  dim(H)=c(N,N,M)
  Hm=aperm(H,c(1,3,2))
  for(i in 1:N){
    Mt[,,i]=t(Hm[,,i])%*%F1%*%Hm[,,i]
    Mt[,,i]=matrix(Mt[,,i]*W[i],nrow = M)
  }
  R=rowSums(Mt,dims = 2)
  # Xmt=matrix(rep(X,each=N,times=1),nrow = N)
  # dim(Xmt)=c(N,N,M)
  # Xm=(aperm(Xmt,c(2,1,3))-Xmt)/sx2
  # Kxm=rep(Kx,times=M)
  # H=matrix(Xm,nrow=N)*matrix(Kxm,nrow=N)
  # Hm=t(H)%*%H
  # #F1t=matrix(rep(t(F1),times=M),ncol = N,byrow = TRUE)
  # #F1tt=t(matrix(rep(F1t),times=M,ncol = N,byrow = TRUE))
  # dim(Hm)=c(N,M,N,M)
  # Hm=aperm(Hm,c(1,3,2,4))
  # dim(Hm)=c(N*N,M,M)
  # F1=rep(F1,times=M*M)
  # dim(F1)=c(N*N,M,M)
  # R=colSums(Hm*F1,dims=1)
  # dim(R)=c(M,M)
  
  V=eigen(R, symmetric = TRUE)
  
}

File: QueueSMC/GKDR.m
function [B_out,V,para_gkdr] = GKDR(G,fix,obstemp,focuspara,band_fix)
%GKDR Summary of this function goes here
%   Detailed explanation goes here
% Bm=fopen('Bmatrix_s.txt','a+');
Np =20000;
Nt=20000;
N=1000; %Np>=4*N according to cv
np=3;
weight_k=1;
c=0.75;
P=size(obstemp,2);
CV_num=1000;
%% generate training data
pri1 = 10*rand(1,Np);                        %theta:3xN
pri2=pri1+10*rand(1,Np);
pri3=1/3.*rand(1,Np);
theta=[pri1' pri2' pri3'];
sampdata=simulationMG1(theta',Np);                    %main sampler 
SumStatQuntSim = quantile(sampdata,P-2,2);
SumStatMaxSim = max(sampdata,[],2);                 %sampdata:Nx50
SumStatMinSim = min(sampdata,[],2);
x = [SumStatMaxSim SumStatMinSim SumStatQuntSim];
tsyE=x';
%tsyE=abs((tsyE-mean(tsyE))./std(tsyE));
% test data
pri1t = 10*rand(1,Nt);                        %theta:3xN
pri2t=pri1t+10*rand(1,Nt);
pri3t=1/3.*rand(1,Nt);
thetat=[pri1t' pri2t' pri3t'];
sampdatat=simulationMG1(thetat',Nt);                    %main sampler 
SumStatQuntSimt = quantile(sampdatat,P-2,2);
SumStatMaxSimt = max(sampdatat,[],2);                 %sampdata:Nx50
SumStatMinSimt = min(sampdatat,[],2);
xt = [SumStatMaxSimt SumStatMinSimt SumStatQuntSimt];
tsyEt=xt';
%tsyEt=abs((tsyEt-mean(tsyEt))./std(tsyEt));
SumStatDat=obstemp;
K = G;

thetas=theta;
[M,tnum]=size(tsyE);

%%
%%%%%%%%%%%%%%%%%%%%%%%%%%
%Simulation     1,no rejection 2.with rejection
%N: number of training data for all parameters(different B)
%1:
starttime=cputime;

%%%calculate distance and random permute
% 
% rtnum=randperm(tnum);
% thetas=thetas(rtnum,:);
% tsyE=tsyE(:,rtnum);

if(fix==0)
    fprintf('Cross validation ...\n');
    fprintf('#sample = %d, dim of X = %d, effective dim = %d\n', N, M,G);
    candx=[0.25 0.5 0.75 1 2];  % candidates for CV
    candy = [0.25 0.5 1 2];
    eps = [0.0001 0.001];%[0.00001];

    % For cross-validation
%     NCV=5;      % Number of cross-validation
%     DEC=10;
%     lx=ceil(N/NCV);
%     ei=cumsum(lx.*ones(1,NCV),2);  
%     si=ei-(lx-1).*ones(1,NCV);
%     ei(NCV)=N;       % si: staring idx, ei: ending idx
    err_tbl=zeros(length(candx)*length(candy)*length(eps), 1);
    rsse_b=zeros(1,Np);

    for in=1:Np                 %(round(N/size_data))
        SumStatSim = tsyE(:,in);    %[SumStatMaxSim SumStatMinSim SumStatQuntSim]';
        rsse_b(in) = minus(SumStatSim,SumStatDat')'*minus(SumStatSim,SumStatDat'); 
    end

    [rsse_t,I]=sort(rsse_b);
    Ym=thetas(I,:);
    %for cv
   % Y_avr=sum(Ym((1:1000),:),1)./1000;
    Xm=tsyE(:,I)';
    Ytt=Ym(1:N,:);
%     Xtt=Xm(1:N,:);
    X=Xm(1:N,:);
    rtt=rsse_t(1:N);
    rsseup=rtt(N);
    W1=1-(rtt/rsseup);
    W1(W1==0)=0.00001;
    W2 = c.*(1.-(rtt./rsseup).^2);
    W3=(1-(rtt./rsseup).^2).^3;
%     Xstd=std(Xtt);
%     Xstd(Xstd==0)=1;
%    Ystd=std(Ytt);
%     X=Xtt./repmat(Xstd,N,1);
%     %Yttm=mean(Y2);
    Y_t=Ytt;%abs((Ytt-mean(Ytt))./std(Ytt));
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    if(weight_k==1)
        W=W1';
    elseif(weight_k==2)
        W=W2';
    elseif(weight_k==3)
        W=W3';
    elseif(weight_k==0)
        W=ones(1,Acc);
    end
    if(focuspara==0)
        Y=Y_t;
    elseif(focuspara==1)
        Y=Y_t(:,1);
    elseif(focuspara==2)
        Y=Y_t(:,2);
    elseif(focuspara==3)
        Y=Y_t(:,3);
    end
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    X=(X-repmat(mean(X),N,1));
    Xstd=std(X);
    Xstd(Xstd==0)=1;
    X=abs(X./repmat(Xstd,N,1));
    Y=(Y-repmat(mean(Y),N,1));
    Ystd=std(Y);
    Y=abs(Y./repmat(Ystd,N,1));

    sgx0=MedianDist(X);   % Basic value for bandwidth for X
    sgy0=MedianDist(Y);   % Basic value for bandwidth for Y 
    Ystd=std(Ytt);
    ridx=randperm(N);  % random order 
    Xr=X(ridx,:);
    Yr=Y(ridx,:);  
    Wr=W(ridx,:); 
    %% parallel validation
    sgx_v=sgx0.*candx;
    sgy_v=sgy0.*candy;
    tt1=length(candx)*length(candy)*length(eps);
    tt2=length(candy)*length(eps);
    for hhh=1:tt1
        opth1=ceil(hhh/tt2);
        sgx=sgx_v(opth1);
        rr1=hhh-(opth1-1)*tt2;
        optk1=ceil(rr1/length(eps));
        sgy=sgy_v(optk1);
        opte1=mod(rr1,length(eps));
        if opte1==0
            opte1=1;
        end 
%     for h=1:length(candx)
%         sgx=sgx0*candx(h);% five different values for CV
%         for k=1:length(candy)
%           sgy=sgy0*candy(k);            
%           for ll=1:length(eps)
            EPS = eps(opte1);
%                 for i=1:NCV
%                     ri=si(i):ei(i);
%                     Xe=Xr; Ye=Yr; We=Wr;
%                     Xe(ri,:)=[];
%                     Ye(ri,:)=[];    % Xe, Ye: trainig sample for CV
%                     We(ri,:)=[];
%                     Xt=Xr(ri,:);
%                     Yt=Yr(ri,:);    % Xt, Yt: test sample for CV
            [B,t,V]=KernelDeriv_for(Xr,Yr,K,sgx,sgy,EPS,Wr);
            
% 
%                     sampdata=simulationMG1(Y_avr',1);                    %main sampler 
%                     SumStatQuntSim_tem = quantile(sampdata,P-2,2);
%                     SumStatMaxSim_tem = max(sampdata,[],2);                 %sampdata:Nx50
%                     SumStatMinSim_tem = min(sampdata,[],2);
%                     x_tem = [SumStatMaxSim_tem SumStatMinSim_tem SumStatQuntSim_tem];

              %  for in=1:10000                 %(round(N/size_data))
%                     SumStatSim = tsyE(1:10000,:)*B;    %[SumStatMaxSim SumStatMinSim SumStatQuntSim]';
%                     x_b=repmat(x_tem,10000,1)*B;
%                     rsse_tmp = minus(SumStatSim,x_b); 
%                     rsse_b1=rsst_tmp'*A*rsse_tmp;
%                 %    end
%                     [~,I]=sort(rsse_b1);
%                     Ym1=thetas(I,:);
%                     Ym2=sum(Ym1(1:1000,:),1)./1000;
%                     dd=abs(Y_avr-Ym2);
            % kNN regression for CV
            Yo=zeros(N,np);
            nnidx=knnsearch(tsyEt'*B,Xr*B, 'K', CV_num, 'NSMethod', 'kdtree'); 
            for jjj=1:N
                indx=nnidx(jjj,:);
                Yo(jjj,:)=sum(thetat(indx',:),1)./CV_num;%Ye
            end
            dd=Yr-Yo;
         %  inx_err=(h-1)*length(candy)*length(eps)+(k-1)*length(eps)+ll;
            err_tbl(hhh)=sum(sum(dd.*dd,1),2);                    
            %5 means divide the data to 5 parts. 
%                 end

%           end
%         end
    end

    %     fprintf('Validation Done\n');

    [c, midx]=min(err_tbl);
    opth=ceil(midx/(length(candy)*length(eps)));
    rr=midx-(opth-1)*length(candy)*length(eps);
    optk=ceil(rr/length(eps));
    opte=mod(rr,length(eps));
    if opte==0
        opte=length(eps);
    end   
    
    %save errortable.mat err_tbl;
%%%%%%%%%%%%%%%%%%%%%%%%
%set up X,Y and weight again as opimum value
    
    rsse_b=zeros(1,Np);
    for in=1:Np  %(round(N/size_data))
        SumStatSim = tsyE(:,in);%[SumStatMaxSim SumStatMinSim SumStatQuntSim]';
        rsse_b(in) = minus(SumStatSim,SumStatDat')'*minus(SumStatSim,SumStatDat'); 
    end
    [rsse_t,I]=sort(rsse_b);
    Ym=thetas(I,:);
    Xm=tsyE(:,I)';
    Ytt=Ym(1:N,:);
    Xtt=Xm(1:N,:);
    rtt=rsse_t(1:N);
    rsseup=rtt(N);
    W1=1-(rtt/rsseup);
    W1(W1==0)=0.00001;
    W2 = c.*(1.-(rtt./rsseup).^2);
    W3=(1-(rtt./rsseup).^2).^3;
    
    Xtt=(Xtt-repmat(mean(Xtt),N,1));
    Xstd=std(Xtt);
    Xstd(Xstd==0)=1;
    X=abs(Xtt./repmat(Xstd,N,1));
    
    Ytt=(Ytt-repmat(mean(Ytt),N,1));
    Ystd=std(Ytt);
    Ystd(Ystd==0)=1;
    Ytt=abs(Ytt./repmat(Ystd,N,1));
   
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    if(weight_k==1)
        W=W1';
    elseif(weight_k==2)
        W=W2';
    elseif(weight_k==3)
        W=W3';
    elseif(weight_k==0)
        W=ones(1,Acc);
    end
    if(focuspara==0)
        Y=Ytt;
    elseif(focuspara==1)
        Y=Ytt(:,1);
    elseif(focuspara==2)
        Y=Ytt(:,2);
    elseif(focuspara==3)
        Y=Ytt(:,3);
    end    
    %%%%%%%%%%%%%%%%%%%%%%%
    sgx=sgx0*candx(opth);
    sgy=sgy0*candy(optk);
    EPS=eps(opte);
    [B,~,V]=KernelDeriv(X,Y,K,sgx,sgy,EPS,W);
    para_gkdr=[sgx sgy EPS];
    B_out=B;
    %   fprintf('NCV=%d, DEC=%d, K=%d\n', NCV, DEC);
    fprintf(' Npop: %d,N: %d',Np,N);
    fprintf(',G= %d sgx=%f,sgy=%f,eps:%f,W:%d,Yfocus:%d\n',K,sgx,sgy,EPS,weight_k,focuspara);
    
%     fprintf('Extimation Error: %f\n', mean(err));
else
%     Y = Y_t(:,1);
    Npcand=Np;
    rsse_b=zeros(1,Npcand);
    for in=1:Npcand  %(round(N/size_data))
        SumStatSim = tsyE(:,in);%[SumStatMaxSim SumStatMinSim SumStatQuntSim]';
        rsse_b(in) = minus(SumStatSim,SumStatDat')'*minus(SumStatSim,SumStatDat'); 
    end
    [rsse_t,I]=sort(rsse_b);
    Ym=thetas(I,:);
    Xm=tsyE(:,I)';
    Ytt=Ym(1:N,:);
    Xtt=Xm(1:N,:);
    rtt=rsse_t(1:N);
    rsseup=rtt(N);
    W1=1-(rtt/rsseup);
    W1(W1==0)=0.00001;
    W2 = c.*(1.-(rtt./rsseup).^2);
    W3=(1-(rtt./rsseup).^2).^3;
    Xtt=(Xtt-repmat(mean(Xtt),N,1));
    Xstd=std(Xtt);
    Xstd(Xstd==0)=1;
    X=abs(Xtt./repmat(Xstd,N,1));
    
    Ytt=(Ytt-repmat(mean(Ytt),N,1));
    Ystd=std(Ytt);
    Ystd(Ystd==0)=1;
    Ytt=abs(Ytt./repmat(Ystd,N,1));
   
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    if(weight_k==1)
        W=W1';
    elseif(weight_k==2)
        W=W2';
    elseif(weight_k==3)
        W=W3';
    elseif(weight_k==0)
        W=ones(1,Acc);
    end
    if(focuspara==0)
        Y=Ytt;
    elseif(focuspara==1)
        Y=Ytt(:,1);
    elseif(focuspara==2)
        Y=Ytt(:,2);
    elseif(focuspara==3)
        Y=Ytt(:,3);
    end    
    %%%%%%%%%%%%%%%%%%%%%%%
    sgx=band_fix(1);
    sgy=band_fix(2);
    EPS=band_fix(3);
    [B,~,V]=KernelDeriv(X,Y,K,sgx,sgy,EPS,W);
    
    B_out=B;
    para_gkdr=band_fix;
    
    %   fprintf('NCV=%d, DEC=%d, K=%d\n', NCV, DEC);
    fprintf(' Npop: %d,N: %d',Npcand,N);
    fprintf(',G= %d sgx=%f,sgy=%f,eps:%f,W:%d,Yfocus:%d\n',K,sgx,sgy,EPS,weight_k,focuspara);
end

% fprintf(Bm,'\nB_out matrix\n%3f\n',B);
% fclose(Bm);
% disp(B_out);
endtime = cputime - starttime;
fprintf('Time = %3.4f\n', endtime);

end

